Metrics Glossary: average precision (first draft)

y27choi opened the pull request:
### Linked issue(s):
[KOL-2580](https://linear.app/kolena/issue/KOL-2580/port-average-precision-doc-to-docskolenaio)

### What change does this PR introduce and why?
* Added the average precision page without linking it to the metrics glossary
* TODO: Diagrams need to be updated to have dark mode support + styling changes (will be done in a separate PR)
* TODO: Should be linked to other pages where "average precision" is referenced. 

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

y27choi committed:
First draft of average precision doc

munkyshi commented:
Noob question: is there a way to view this locally since it isn't linked?

gordonhart commented:
@munkyshi visit the URL (the file path, sans `.md`, from `./docs`): [localhost:8999/metrics/average-precision](http://localhost:8999/metrics/average-precision/)

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,250 @@
+# Average Precision
```
Add search exclusion? Looks like the plan is to land this with the temp diagrams and follow up with the production diagrams + links to this guide on the sidebar and index page.

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,250 @@
+all thresholds.
+
+AP is commonly used to evaluate the performance of object detection and information retrieval workflows. This metric
+(or an aggregated version of it called **mean average precision** (mAP)) is the primary metric used across popular
```
We should add a subsection for mAP — just a simple description like this, along with something saying that it's usually the macro-average across all classes.

munkyshi commented:
> @munkyshi visit the URL (the file path, sans `.md`, from `./docs`): [localhost:8999/metrics/average-precision](http://localhost:8999/metrics/average-precision/)

🤦 I was trying to load the page from `trunk` instead of this branch

munkyshi commented on docs/assets/images/metrics-ap-ir-example.png:
This image doesn't toggle well with dark mode.

y27choi replied:
These are placeholder diagrams.  Will follow up with another PR for diagram updates.

y27choi committed:
Added map section

y27choi committed:
Merge remote-tracking branch 'origin/trunk' into yoohee/metrics-glosary-ap

y27choi committed:
Merge remote-tracking branch 'origin/trunk' into yoohee/metrics-glosary-ap

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,261 @@
+
+Average precision (AP) summarizes a precision-recall (PR) curve into a single value representing the average of all
+precisions. It is generally understood as the approximation of the area under the PR curve. This metric ranges between
+0 to 1, where a perfect model with a [precision](./precision.md) and [recall](./recall.md) of 1 scores 1.
```
Wording here is awk — suggestion:

> AP ranges between 0 and 1, where a perfect model has precision, recall, and AP scores of 1.

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,261 @@
+Unlike metrics like [precision](./precision.md), [recall](./recall.md), and [F<sub>1</sub>-score](./f1-score.md), which
+are threshold-dependent where a confidence threshold value must be defined to compute them, AP is a key performance
+metric that removes the dependency of selecting one confidence threshold value and measures a model's performance across
+all thresholds.
```
We should include "threshold-independent" bolded somewhere. This feels like an important piece of information that should pop

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,261 @@
+based on the [Intersection over Union (IoU)](./iou.md) scores. Let’s look at each inference bounding box and sort them
+by their confidence score in descending order.
+
+| Inference | Confidence ↓ | TP/FP | cumsum(TP) | cumsum(FP) | Precision | Recall |
```
Nit: wrap confidence and sort arrow in `<nobr>`

y27choi replied:
Didn't know you can do that.  TIL!

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,261 @@
+
+![object detection example - PR curve](../assets/images/metrics-ap-od-pr.png)
+
+Notice the zigzag pattern, called “**wiggles**” — the precision goes down with FPs and goes up again with TPs as the
```
Is "wiggles" the technical term?

y27choi replied:
Funny enough, it is a term that was used in PASCAL VOC paper. I reworded it to make it less official.

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,261 @@
+\text{AP} = \frac {(\frac 1 1 + \frac 0 2 + \frac 0 3 + \frac 2 4 + \frac 3 5 + \frac 0 6 + ... + 0)} 3 = 0.7
+$$
+
+### mean Average Precision (mAP)
```
We should capitalize `Mean`

gordonhart commented on docs/metrics/average-precision.md:
```diff
@@ -0,0 +1,261 @@
+this metric is very popular and commonly used in object detection and information retrieval workflows, it has some
+limitations. Let's make sure to understand these limitations before using the metric to compare your models.
+
+1. AP is often overestimated.
```
I like these limitations a lot. Can we make them pop more? Maybe bolding the header, or making each into a callout (using [`info`](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#type:info), maybe)?

y27choi replied:
Tried the callout.  Let me know if you think this is too much
![Uploading Screen Shot 2023-07-11 at 4.32.08 PM.png…]()

gordonhart replied:
What do you think about making them collapsed (and `info`) by default? Up to you — reverting to a numbered list would also be fine.

![Screen Shot 2023-07-11 at 4 37 09 PM](https://github.com/kolenaIO/kolena/assets/9648886/3a55cc70-8423-4a45-96be-10ca61e4fe40)

y27choi replied:
I'm good with the collapsible info.  I will make this change in a separate PR quickly.

gordonhart reviewed (approved):
Bulk of the content looks great. Nice amount of detail here. Approving with a few small comments.

Is @pvanburen working on these diagrams already?

y27choi committed:
Addressed PR comments

y27choi commented:
Thanks for your thorough review, @gordonhart.  I addressed all your comments.  
And yes @pvanburen is working on the AP diagrams now.  She will open another PR for diagram update.
